// rxjs/src/Subject.js
"use strict";
const Observable_1 = require('./Observable');
const Subscriber_1 = require('./Subscriber');
const Subscription_1 = require('./Subscription');
const SubjectSubscription_1 = require('./SubjectSubscription');
const rxSubscriber_1 = require('./symbol/rxSubscriber');
const throwError_1 = require('./util/throwError');
const ObjectUnsubscribedError_1 = require('./util/ObjectUnsubscribedError');
/**
 * @class Subject<T>
 */
class Subject extends Observable_1.Observable {
    constructor(destination, source) {
        super();
        this.destination = destination;
        this.source = source;
        this.observers = [];
        this.isUnsubscribed = false;
        this.isStopped = false;
        this.hasErrored = false;
        this.dispatching = false;
        this.hasCompleted = false;
        this.source = source;
    }
    lift(operator) {
        const subject = new Subject(this.destination || this, this);
        subject.operator = operator;
        return subject;
    }
    add(subscription) {
        return Subscription_1.Subscription.prototype.add.call(this, subscription);
    }
    remove(subscription) {
        Subscription_1.Subscription.prototype.remove.call(this, subscription);
    }
    unsubscribe() {
        Subscription_1.Subscription.prototype.unsubscribe.call(this);
    }
    _subscribe(subscriber) {
        if (this.source) {
            return this.source.subscribe(subscriber);
        }
        else {
            if (subscriber.isUnsubscribed) {
                return;
            }
            else if (this.hasErrored) {
                return subscriber.error(this.errorValue);
            }
            else if (this.hasCompleted) {
                return subscriber.complete();
            }
            this.throwIfUnsubscribed();
            const subscription = new SubjectSubscription_1.SubjectSubscription(this, subscriber);
            this.observers.push(subscriber);
            return subscription;
        }
    }
    _unsubscribe() {
        this.source = null;
        this.isStopped = true;
        this.observers = null;
        this.destination = null;
    }
    next(value) {
        this.throwIfUnsubscribed();
        if (this.isStopped) {
            return;
        }
        this.dispatching = true;
        this._next(value);
        this.dispatching = false;
        if (this.hasErrored) {
            this._error(this.errorValue);
        }
        else if (this.hasCompleted) {
            this._complete();
        }
    }
    error(err) {
        this.throwIfUnsubscribed();
        if (this.isStopped) {
            return;
        }
        this.isStopped = true;
        this.hasErrored = true;
        this.errorValue = err;
        if (this.dispatching) {
            return;
        }
        this._error(err);
    }
    complete() {
        this.throwIfUnsubscribed();
        if (this.isStopped) {
            return;
        }
        this.isStopped = true;
        this.hasCompleted = true;
        if (this.dispatching) {
            return;
        }
        this._complete();
    }
    asObservable() {
        const observable = new SubjectObservable(this);
        return observable;
    }
    _next(value) {
        if (this.destination) {
            this.destination.next(value);
        }
        else {
            this._finalNext(value);
        }
    }
    _finalNext(value) {
        let index = -1;
        const observers = this.observers.slice(0);
        const len = observers.length;
        while (++index < len) {
            observers[index].next(value);
        }
    }
    _error(err) {
        if (this.destination) {
            this.destination.error(err);
        }
        else {
            this._finalError(err);
        }
    }
    _finalError(err) {
        let index = -1;
        const observers = this.observers;
        // optimization to block our SubjectSubscriptions from
        // splicing themselves out of the observers list one by one.
        this.observers = null;
        this.isUnsubscribed = true;
        if (observers) {
            const len = observers.length;
            while (++index < len) {
                observers[index].error(err);
            }
        }
        this.isUnsubscribed = false;
        this.unsubscribe();
    }
    _complete() {
        if (this.destination) {
            this.destination.complete();
        }
        else {
            this._finalComplete();
        }
    }
    _finalComplete() {
        let index = -1;
        const observers = this.observers;
        // optimization to block our SubjectSubscriptions from
        // splicing themselves out of the observers list one by one.
        this.observers = null;
        this.isUnsubscribed = true;
        if (observers) {
            const len = observers.length;
            while (++index < len) {
                observers[index].complete();
            }
        }
        this.isUnsubscribed = false;
        this.unsubscribe();
    }
    throwIfUnsubscribed() {
        if (this.isUnsubscribed) {
            throwError_1.throwError(new ObjectUnsubscribedError_1.ObjectUnsubscribedError());
        }
    }
    [rxSubscriber_1.$$rxSubscriber]() {
        return new Subscriber_1.Subscriber(this);
    }
}
Subject.create = (destination, source) => {
    return new Subject(destination, source);
};
exports.Subject = Subject;
/**
 * We need this JSDoc comment for affecting ESDoc.
 * @ignore
 * @extends {Ignored}
 */
class SubjectObservable extends Observable_1.Observable {
    constructor(source) {
        super();
        this.source = source;
    }
}
